### PR TITLE
[GR-47106] Return to separate CUs per class to improve gdb startup time

### PR DESCRIPTION
**Pull request by adinn**
This patch restores the DWARF debug info generator to generating a compilation unit (CU) per Java class. This helps alleviate the performance problem identified in issue  #6936.

### CODE REVIEW

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java [hunk 1]
```diff
@@ -313,8 +364,99 @@ public int lowpc() {
      *
      * @return the highest code section offset for compiled method code belonging to this class
      */
+    @SuppressWarnings("unused")
```
Comment (fniephaus) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:367]: Not sure a public method really needs this.

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java [hunk 2]
```diff
@@ -313,8 +364,99 @@ public int lowpc() {
...
+     * Add a directory to the list of firectories referenced from info associated with this class.
+     * 
+     * @param dirEntry The directory to be added.
+     */
+    public void includeDir(DirEntry dirEntry) {
+        if (dirs == null) {
+            dirs = new ArrayList<>();
+        }
+        assert !dirs.contains(dirEntry);
+        // cannot add dirs after index has been created
```
Comment (fniephaus) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:398]: should this be the assertion description?

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java [hunk 3]
```diff
@@ -313,8 +364,99 @@ public int lowpc() {
...
+     * Populate the file and directory indexes that track positions in the file and dir tables for
+     * this class's line info section.
+     */
+    public void buildFileAndDirIndexes() {
+        // this is a one-off operation
+        assert fileIndex == null;
+        assert dirIndex == null;
+        if (files == null) {
+            // should not have any dirs if we have no files
+            assert dirs == null;
```
Comment (fniephaus) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:413]: same here: turn comment into assertion description?

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfDebugInfo.java [hunk 4]
```diff
@@ -70,54 +71,56 @@ public class DwarfDebugInfo extends DebugInfoBase {
      */
     @SuppressWarnings("unused") public static final int DW_ABBREV_CODE_null = 0;
     /* Level 0 DIEs. */
-    public static final int DW_ABBREV_CODE_class_unit = 1;
+    public static final int DW_ABBREV_CODE_builtin_unit = 1;
```
Comment (fniephaus) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfDebugInfo.java:74]: should we turn this into an enum or use a control variable to initialize the constants? Seems rather annoying that we have to bump each of them individually.

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java (hunk 1 above)
Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:367]: Yes, removed.

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java (hunk 2 above)
Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:398]: I have updated the asserts to include the relevant comment as the assert message.

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfDebugInfo.java (hunk 4 above)
Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfDebugInfo.java:74]: Yeah that would probably be a good idea. However, I would prefer to defer that to a separate patch. I think this one may need backporting.

**Comment by adinn** [comment]
Thanks for the review Fabio!

**Comment by fniephaus** [comment]
>Thanks for the review Fabio!

Of course, thanks for addressing. While we're waiting for Paul's review, I've started the integration process already and am running the internal gates on this.

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DirEntry.java [hunk 5]
```diff
@@ -37,11 +37,9 @@
  */
 public class DirEntry {
```
Comment (fniephaus) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DirEntry.java:38]: Do we really still need `DirEntry` or should we simplify this and just use `Path`?

Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DirEntry.java:38]: Yes, I guess so. Although I like the fact that using this class clearly demarcates the Path as a directory and not, say, a file.

n.b. It's certainly not costing much to wrap the path. The Quarkus app that manifested the performance problem has 500 dirs with full debug info. The number is much less if `TrackNodeSourcePosition` or `SourceLevelDebug` is omitted.

Comment (fniephaus) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DirEntry.java:38]: Sure, if you prefer a clear separation between dir and file, keep the wrapper. If you want to optimize SLOC, you can even use a record ;)

**Review by zakkak — APPROVED** [review]
LGTM.

As mentioned in https://github.com/Karm/mandrel-integration-tests/issues/160#issuecomment-1621570077 it still feels slow but it's about 3x faster on my machine

**Comment by olpaw** [comment]
@adinn please fixup the commits
```
* fd4313e26ef - (6 days ago) more style fixes - Andrew Dinn
* 2623562c669 - (6 days ago) style fixes - Andrew Dinn
* 8d5750d2fc5 - (6 days ago) style fixes - Andrew Dinn
```
into a single commit. Also it would be nice if the commit message would all start uppercase.

**Comment by adinn** [comment]
> @adinn please fixup the commits

Done.

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DebugInfoBase.java [hunk 6]
```diff
@@ -741,4 +728,51 @@ public boolean isHubClassEntry(ClassEntry classEntry) {
...
+                if (!subRange.isLeaf()) {
+                    checkInclude(classEntry, subRange.getFileEntry(), visitedFiles, visitedDirs);
+                }
+            });
+        });
+        // now all files and dirs are known build an index for them
+        classEntry.buildFileAndDirIndexes();
+    }
+
+    private static void checkInclude(ClassEntry classEntry, FileEntry fileEntry, EconomicSet<FileEntry> visitedFiles, EconomicSet<DirEntry> visitedDirs) {
```
Comment (olpaw) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DebugInfoBase.java:767]: Calling a method that adds to collections `checkInclude` is confusing.

Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DebugInfoBase.java:767]: I have changed the name to `ensureIncluded` and added Javadoc explaining what it does.

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java [hunk 7]
```diff
@@ -317,4 +368,90 @@ public int hipc() {
...
+        files.add(file);
+    }
+
+    /**
+     * Add a directory to the list of firectories referenced from info associated with this class.
+     * 
+     * @param dirEntry The directory to be added.
+     */
+    public void includeDir(DirEntry dirEntry) {
+        if (dirs == null) {
```
[the remark this answers is not in the document]
Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:392]: > Does it occur often that a ClassEntry's  `files` and `dirs` fields are unused

I believed this happened frequently enough that on-demand create would save a useful amount of space. However, I checked and it is highly abnormal to have zero files or dirs. So, I think it is better to make these fields final and populate them with a list at at ClassEntry create. A patch is on its way.

[the remark this answers is not in the document]
Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java]: Fix pushed

**Review by olpaw — APPROVED** [review]

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java [hunk 8]
```diff
@@ -90,12 +90,12 @@ public class ClassEntry extends StructureTypeEntry {
      * A list of all files referenced from info associated with this class, including info detailing
      * inline method ranges.
      */
-    private ArrayList<FileEntry> files;
+    final private ArrayList<FileEntry> files;
```
Comment (olpaw) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:93]: Please use `private final` here and below to make checkstyle happy

Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java]: Oops, sorry about that. Done.

**Comment by fniephaus** [comment]
Hi @adinn, we failed to integrate this due to a few memory-constraint jobs that are now reliably failing with your changes, either because they're running out of memory or because they're killed by the deadlock watchdog (very likely also because the GC is very busy and slows down the creation of debuginfo). Before we increase the memory limits of these jobs, could you maybe check whether there's something that can be done to bring the memory usage back down? Here's an example stack trace from a watchdog-killed job:

```
=== Image generator watchdog detected no activity. This can be a sign of a deadlock during image building. Dumping all stack traces. Current time: Thu Jul 13 11:28:43 UTC 2023
"main" Id=1 in RUNNABLE
    at java.base@21/java.util.ArrayList.spliterator(ArrayList.java:1613)
    at java.base@21/java.util.Collection.stream(Collection.java:750)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.debugentry.ClassEntry.compiledEntries(ClassEntry.java:249)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.elf.dwarf.DwarfARangesSectionImpl.lambda$writeContent$2(DwarfARangesSectionImpl.java:148)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.elf.dwarf.DwarfARangesSectionImpl$$Lambda/0x00000001021af108.accept(Unknown Source)
    at java.base@21/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
    at java.base@21/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
    at java.base@21/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1707)
    at java.base@21/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
    at java.base@21/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
    at java.base@21/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
    at java.base@21/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
    at java.base@21/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
    at java.base@21/java.util.stream.ReferencePipeline.forEachOrdered(ReferencePipeline.java:601)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.elf.dwarf.DwarfARangesSectionImpl.writeContent(DwarfARangesSectionImpl.java:144)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.elf.dwarf.DwarfSectionImpl$$Lambda/0x00000001021ae3b8.accept(Unknown Source)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.ObjectFile.debugContext(ObjectFile.java:1833)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.elf.dwarf.DwarfSectionImpl.getOrDecideContent(DwarfSectionImpl.java:647)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.elf.dwarf.DwarfARangesSectionImpl.getOrDecideContent(DwarfARangesSectionImpl.java:131)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.elf.ELFUserDefinedSection.getOrDecideContent(ELFUserDefinedSection.java:110)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.ObjectFile.bake(ObjectFile.java:1625)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.ObjectFile.write(ObjectFile.java:1282)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.ObjectFile.lambda$write$5(ObjectFile.java:1274)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.ObjectFile$$Lambda/0x00000001021a8d58.run(Unknown Source)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.ObjectFile.withDebugContext(ObjectFile.java:1814)
    at app/org.graalvm.nativeimage.objectfile/com.oracle.objectfile.ObjectFile.write(ObjectFile.java:1273)
    at app/org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImage.write(NativeImage.java:166)
    at app/org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:98)
    at app/org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.SharedLibraryImageViaCC.write(SharedLibraryImageViaCC.java:57)
```

**Comment by adinn** [comment]
@fniephaus The latest (espresso) gate failure appears to be unrelated to this patch. I think this is ready to be resubmitted to the internal gate.

**Comment by fniephaus** [comment]
> I looked at what was driving the high memory use in the debug info generator and have managed to remove a lot of the overhead.

Fantastic, thanks a lot, @adinn! Let me run the gates again and see whether it passes our benchmarking infra. :)

And special thanks for the detailed analysis report. I always enjoy reading them!


> The latest (espresso) gate failure appears to be unrelated to this patch. I think this is ready to be resubmitted to the internal gate.

No need to worry about it. We're already working on fixing that.

**Comment by adinn** [comment]
> And special thanks for the detailed analysis report. I always enjoy reading them!

Well, indeed, much as I do when it is 6 months down the line and I have near zero recollection of what changes I made or why I made them ;-). All the same, your words of appreciation are gratefully received.

**Comment by fniephaus** [comment]
... and we're past the failing gates, yay! Thanks for fixing...getting this merged now (see #6984).

**Pull request merged**